### ISSUE TITLE
Tool-call decoding is ~4x slower than plain decode: grammar mask computation on the host path

### ISSUE THREAD

**Issue by TheTom** [Avarok-Cybersecurity/atlas#910]
## Summary

On a single H100 (CC 9.0, driver 610.57, `sm_90a` build at `f66a262`), Qwen3.6-35B-A3B-FP8, temperature 0, thinking off: plain chat decodes at ~245 tok/s, but any request that carries a tool schema decodes at **~56 tok/s** for the duration of the tool call, and the first tool call after a cold start waits **~4.9 s** before the first token. The same requests against a second engine on the same GPU and checkpoint produce the same text, with the tool call finished in 0.29 s. Correctness is fine (12/12 valid calls across `tool_choice` auto/required/named × streaming/blocking); this is purely a latency defect in the constrained-decoding path.

## Measurements (all on the H100, evidence retained on the campaign docs branch under `docs/campaigns/hopper-atlas-vs-vllm-2026-09/evidence/rental-h100-20260905/tool-call-audit/`)

| request | first tool token | finish | tokens/s during tool response |
|---|---|---|---|
| tool call, cold (first schema seen) | 4.94 s | 5.66 s | 54 |
| tool call, warm (same schema again) | 0.115 s | 0.80 s (~41 tokens) | 56 |
| plain chat, same server | 0.077 s | 0.176 s (~20 tokens) | ~245 |

`ATLAS_DECODE_TIMING=1` over the last 100 host-path tokens of warm tool calls:

```
copy+fwd-wait = 4.11 ms/tok   sample(248k host) = 13.78 ms/tok
```

Any grammar-bearing sequence forces the host sampling path (`crates/spark-server/src/scheduler/decode_logits_step.rs:132`); plain greedy rows stay on GPU argmax. The 13.78 ms is not the arithmetic: a native micro-benchmark of the host loops (BF16 expand, mask apply, top-two scan, FP32 copy + greedy over the 248 K vocab) totals ~0.24 ms/token. It is the **grammar mask computation**. A CPU replay of the exact 40-token response shows the fill cost concentrated on four positions:

| position | token | fill |
|---|---|---|
| 13, 14 | `\n`, `Re` (start of the `"Reykjavik"` value) | ~68 ms each |
| 28, 29 | `\n`, `3` (start of the `days` value) | ~68 ms each |
| all 36 others | | ~0.1 ms each |

272 of 273 ms come from the two free-form JSON value slots, where the grammar admits nearly the whole vocabulary and the next-token mask is computed on demand instead of being served from a compiled mask.

## Cold start

`GrammarState::build` calls `compile_top_k_masks(512)` synchronously before prefill and computes the selected states serially (`compiled_grammar.rs` ~363, `get_or_compute_mask`). Parallelising that compile only brought 2.18 s → 1.72 s in a release build, so it is not the main fix. The first request for a schema in a server's lifetime pays ~4.9 s; the compiled masks are not persisted across restarts.

## Already fixed on #895

- `d1322aa1` removed a duplicate rendering of the tool schema (Atlas prepended its own tool-instruction block and also passed `tools` to the checkpoint's native template): 1,135 → 298 prompt tokens for the same request.
- `55784398` makes `time_to_first_token_ms` include grammar preparation; before, the reported TTFT for a 4.96 s request was 460 ms.

Neither changes the per-token or cold costs above.

## Suggested directions

1. Treat near-full masks as a complement (allow-all minus a small deny set) or precompute masks for the "any string / any number" states so value slots do not trigger a full-vocabulary fill per token.
2. Persist or precompile grammars per tool-schema hash so the first agent request of a session does not pay the compile.
3. For `tool_choice=auto`, consider not constraining decoding at all and parsing the emitted call instead (the model produced a correct call in every run here without needing the mask); reserve the grammar for `required`/named where the format must be forced.
4. Longer term, apply the bitmask and argmax on device so grammar-bearing rows do not fall off the GPU path (the 4.11 ms/token transfer + wait would also disappear).

## Reproduction

Serve Qwen3.6-35B-A3B-FP8 (any Hopper or GB10 build), send the campaign coherency gate's weather tool request (`bench/hopper_ab/coherency_gate.py`, the `toolcall_ok` probe) twice with streaming on, and compare first-tool-chunk and finish times against a plain 20-token chat request; run with `ATLAS_DECODE_TIMING=1` to see the host-path split.

**Comment by TheTom** [comment]
I am working on this during the rental. The H100 trace confirms 13.78 ms/token host sampling versus 4.11 ms/token copy **plus GPU forward wait**, and the corrected native tool requests pass12/12. [Preserved runtime/CPU evidence](https://github.com/TheTom/atlas/blob/9fb23e4/docs/campaigns/hopper-atlas-vs-vllm-2026-09/RENTAL-H100-TOOL-CALLS.md).

Two attribution details matter for the fix:

- The ~68 ms value-entry fills are from an optimized **Spark1 CPU replay** of the exact native grammar/tokenizer, not an H100-host timing. It averages6.83 ms over40 content tokens; accept_token is about1 microsecond. The H100 host is separately measured by the13.78 ms aggregate. The non-fill loop microbenchmark was also on a different CPU; its pinned Rust1.93.1 rerun including scratch-buffer operations is about0.60 ms/token.
- The4.11 ms includes GPU computation wait; moving mask/argmax onto the GPU would not make that whole interval disappear. Likewise, valid tool output here does not establish that unconstrained auto decoding is equivalent: the installed vLLM0.28 parser supports qwen_3_coder structural tags for auto/required/named and defaults VLLM_ENFORCE_STRICT_TOOL_CALLING to true. We are preserving constraints while fixing their cost.

Current work: reproduce/isolate the four value-entry mask spikes, then test an exactness-preserving matcher optimization. The oracle compares **every mask at every position**, accepted and malformed continuations, and rollback. Bounded parallel cold prewarm already passed824 xgrammar tests and complete-mask replay; its release median improvement on Spark1 is only2.186→1.723s for required, so it is a secondary change. A red-first warm-cache test now also prevents spawning workers when all selected masks are already cached.

No decode speedup is claimed yet. I will post the actual H100 before/after build identities and timings once the warm fix passes the CPU oracle and live tool matrix. Avoid parallel edits to xgrammar matcher/compiler while these two scoped agents finish; fixes will land on #895 with a new perf-tree tag.

**Comment by TheTom** [comment]
The warm value-entry fix now has actual H100 confirmation, using the same Qwen3.6 FP8 checkpoint, native tool request bytes, recipe flags and recorded `ATLAS_DECODE_TIMING=1`. Source `ff7518071b1cbab18c5436ed4554e9a121b882b2`, binary SHA256 `d292e71b416ca93c694eb0ac0754d018549e2789238e5d25f5c9ee5652341c78`. The tested commits are currently local/on the rental and will be pushed with the remaining cold-path repair.

| Diagnostic | Original f66a262 | Fixed, two fresh processes |
|---|---:|---:|
| Host sampling |13.78 ms/token|1.55–1.68 ms/token|
| GPU forward wait plus copy |4.11 ms/token|4.16–4.21 ms/token|
| Tool decode, engine-reported |about56 tokens/s|about162–177 tokens/s|
| First repeat: request to complete call |0.798 s|0.317 /0.321 s|
| First schema: request to first tool delta |4.824 s|6.567 /5.279 s|

The twelve auto/required/named × streaming/blocking × two-repetition typed tool checks pass. All ten plain/tool timing probes across the two fresh processes pass. Original/factored production grammar masks and rollback match byte-for-byte on the pinned tokenizer; the combined CPU server suite passes2456 tests, and workspace doctests/rustdoc pass.

**Warm decode is materially better; cold latency remains open.** Factoring the native parameter-value production lets the existing scanner handle value entry, while keeping constraints active. The cold follow-up found that mask interval preparation clones the full sorted vocabulary for every state—about30million token-buffer allocations over123 states. A red-first borrowed-input fix is being tested now; no extra cold speedup is claimed yet. A duplicate-cache-computation hypothesis was checked and rejected (123 distinct keys, zero duplicate computes).

Raw files are exported on the Mac under `rental-live/single-h100/remote-results/diagnostic.atlas.grammarfix01/` and `grammarfix02/`; the documentation publication follows with checksums. Both owned sessions stopped, and subsequent fresh GPU tenant queries were empty. Immediate teardown snapshots briefly retained the exiting owned PID as `[No data]`; this is preserved, not silently rewritten. These are diagnostic before/after measurements, not certified ladder scores; the independent word-reversal coherency failure remains.

**Comment by TheTom** [comment]
The grammar fixes are now pushed on PR #895 through `fd65fb85a09e9ff3d81c36fcbceffff4bdc38269` (new perf-tree tag `gate-tip-fd65fb8`). Constraints remain enabled.

The final fresh H100 Qwen3.6 diagnostic used source `37228e85a72230e3857dbbb6c8d3286793793887`, executable SHA256 `e2c1a9777b3ad4e38d58314d2bd30a0f8c37b3d1dd96d3d0a10cf6e9de062b18`: first-schema first tool delta **3.379 s**, complete call **3.573 s**; first repeated call first delta **0.106 s**, complete **0.306 s**. Engine-reported tool decode was **169–179 tokens/s**; all five timing probes and twelve typed tool cases passed. The original diagnostic was about56 tokens/s and0.798 s to complete the first repeated call.

The final engine keeps its historical single-worker compiler default. Borrowing immutable tokenizer inputs removes vocabulary-sized allocation growth (red12,356→24,644 allocations versus green58→58); four-worker H100 cold results remained5.28–6.57 s, so parallel execution is not enabled by default. The 3.379 s cold observation is one fresh-process result, not a distribution or a claim that cold latency matches vLLM. Subsequent fully cached prompts still show about0.31 s of internal prefill and remain a separate open finding.

Final combined CPU gates: kernels75, model645, server2,456; xgrammar826 unit tests plus the allocation integration test. Optional native-tokenizer entries without their input were not counted as new native replay evidence. Full-mask/rollback equivalence and earlier actual-tokenizer receipts are preserved, along with all unsuccessful H100 variants. Documentation/raw evidence publication follows on the campaign docs branch; no frozen ladder or benchmark threshold was changed.
